ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -1245,4 +1247,119 @@ int ompi_coll_base_allreduce_intra_redscat_allgather(
+ */
+int ompi_coll_base_allreduce_intra_allgather_reduce(const void *sbuf, void *rbuf, int count,
+                                                    struct ompi_datatype_t *dtype,
+                                                    struct ompi_op_t *op,
+                                                    struct ompi_communicator_t *comm,
+                                                    mca_coll_base_module_t *module)
+{
+    char *send_buf = sbuf;
+    int comm_size = ompi_comm_size(comm);
+    int err = MPI_SUCCESS;
+    int rank = ompi_comm_rank(comm);
+    bool commutative = ompi_op_is_commute(op);
+    ompi_request_t **reqs;
+
+    if (sbuf == MPI_IN_PLACE) {
+        send_buf = rbuf;
+    }
+
+    /* Allocate a large-enough buffer to receive from everyone else */
+    char *tmp_buf = NULL, *tmp_buf_raw = NULL, *tmp_recv = NULL;
+    ptrdiff_t lb, extent, dsize, gap = 0;
+    ompi_datatype_get_extent(dtype, &lb, &extent);
+    dsize = opal_datatype_span(&dtype->super, count * comm_size, &gap);
+    tmp_buf_raw = (char *) malloc(dsize);
+    if (NULL == tmp_buf_raw) {
+        return OMPI_ERR_OUT_OF_RESOURCE;
+    }
+
+    if (commutative) {
+        ompi_datatype_copy_content_same_ddt(dtype, count, (char *) rbuf, (char *) send_buf);
+    }
+
+    tmp_buf = tmp_buf_raw - gap;
+
+    /* Requests for send to AND receive from everyone else */
+    int reqs_needed = (comm_size - 1) * 2;
+    reqs = ompi_coll_base_comm_get_reqs(module->base_data, reqs_needed);
+
+    ptrdiff_t incr = extent * (ptrdiff_t) count;
+    tmp_recv = (char *) tmp_buf;
+
+    int req_index = 0;
+    for (int peer_rank = 0; peer_rank < comm_size; peer_rank++) {
+        tmp_recv = tmp_buf + (peer_rank * incr);
+        if (peer_rank == rank) {
+            /* Skip send/recv from self - later we copy the data to recv buffer */
+            continue;
+        }
+
+        err = MCA_PML_CALL(irecv(tmp_recv, count, dtype, peer_rank, MCA_COLL_BASE_TAG_ALLREDUCE,
+                                 comm, &reqs[req_index++]));
+        if (MPI_SUCCESS != err) {
+            goto err_hndl;
+        }
+
+        err = MCA_PML_CALL(isend(send_buf, count, dtype, peer_rank, MCA_COLL_BASE_TAG_ALLREDUCE,
+                                 MCA_PML_BASE_SEND_STANDARD, comm, &reqs[req_index++]));
+        if (MPI_SUCCESS != err) {
+            goto err_hndl;
+        }
+    }
+
+    /* Copy my data to recv buffer */
+    memcpy(tmp_buf + (rank * incr), send_buf, incr);
+
+    err = ompi_request_wait_all(req_index, reqs, MPI_STATUSES_IGNORE);
+
+    char *inbuf;
+    for (int peer_rank = 0; peer_rank < comm_size; peer_rank++) {
+        inbuf = tmp_buf + (peer_rank * incr);
+        if (0 == peer_rank && !commutative) {
+            /* Sort the data buffer for non-commutative operations */
+            memcpy(rbuf, inbuf, incr);
+            continue;
+        }
+        ompi_op_reduce(op, (void *) inbuf, rbuf, count, dtype);
+    }
```

Comment:
I'm confused about this part.  It seems that the if statement can only be hit on the first iteration of the loop (when peer_rank is 0) and its only effect is to prime the loop to start the reduction with rbuf, however isn't this required for both commutative and non-commutative operations?  Also, this isn't sorting the data buffer, you've already collected the data in sorted order above, so I think the comment is misleading.

Reply:
This is to ensure the order of reduction is the same across ranks. For example:
- Rank 0 sends 1.11
- Rank 1 sends 2.22
- Rank 3 sends 3.33
- ...

If we want to do `MPI_SUM` which is non-commutative, we need to ensure the order `1.11 + 2.22 + 3.33 + ...`

Without L1333, we have something different:
- Rank 0 does `1.11 + 2.22 + 3.33 + ...`
- Rank 1 does `2.22 + 1.11 + 3.33 + ...`
- Rank 2 does `3.33 + 1.11 + 2.22 + ...`
- ...

This means every rank could potentially get different result.

But I sense there might be a more elegant/efficient way to ensure this. I'm open to better ideas.

Reply:
But after wait_all and because all your recv's put their results in the buffer in the correct order, there is no ordering left to resolve...  The buffer is ordered by peer rank already, or did I miss something?

Reply:
The problem is on `ompi_op_reduce(op, inbuf, rbuf, count, dtype)`, i.e.
>  Perform a reduction operation with count elements of type dtype in
the buffers source and target.  The target buffer obtains the
result (i.e., the original values in the target buffer are reduced
with the values in the source buffer and the result is stored in
the target buffer).

On L1292 `rbuf` is populated with send buff data, so this means the reduction order will be different for each rank.